File: crates/lanekeep-rules/rules/no-assertionless-test.ts
import { defineRule } from 'lanekeep'

/**
 * A test that asserts nothing passes forever and covers nothing.
 *
 * Agents pad coverage on request — a body that calls the subject and checks nothing is the
 * cheapest way to make a coverage number move — so this fires often and early in
 * agent-written code. One rule, four language families: what is per-language is how a test
 * is recognized and what counts as asserting; the judgment is the same everywhere.
 *
 * | language | a test is | asserts by default |
 * | --- | --- | --- |
 * | typescript/tsx | an `it(...)`/`test(...)` call with a block-bodied callback | `expect*`, `assert*` |
 * | python | a `def test*` function, methods included | the `assert` statement, `self.assert*`, `self.fail`, `pytest.raises` |
 * | go | `func Test*` taking `*testing.T` | `t.Error*`, `t.Fatal*`, `t.Fail*`, `assert.*`, `require.*` |
 * | rust | a `fn` under `#[test]` or a `::test` attribute path | `assert*!`, `debug_assert*!`, `panic!` |
 *
 * Two exemptions are correctness rather than convenience: a go test that calls `t.Skip*`
 * and a rust test under `#[should_panic]` legitimately assert nothing.
 *
 * Vocabulary entries are matched as *prefixes* of the normalized callee (whitespace
 * stripped, `?.` folded to `.`), so `t.Error` covers `t.Errorf` and `self.assert` covers
 * every `self.assert*` method. The rule does not chase helpers: an assertion inside a
 * function the test calls is invisible here, which is the same limit `expect-expect` has —
 * name such helpers in `allowHelpers` and they count as asserting in every language.
 *
 * Known limits, deliberate for v1: go's receiver is matched by its conventional name (a
 * `func TestX(tt *testing.T)` calling `tt.Error` needs `assertions: { go: ['tt.'] }`), and
 * table-driven tests whose assertion lives in a loop body are covered only because the
 * loop is still inside the test's block.
 *
 * @example
 * ```ts
 * import noAssertionlessTest from 'lanekeep/no-assertionless-test'
 *
 * export default defineConfig({
 *   rules: [
 *     noAssertionlessTest({
 *       tests: ['tests/**', 'src/**'],
 *       assertions: { go: ['suite.'] },
 *       allowHelpers: ['expectValidResponse'],
 *     }),
 *   ],
 * })
 * ```
 */
export default function noAssertionlessTest(options) {
  // The ignored-options trap: options reach a rule only by being closed over.
  const tests = options?.tests
  const extra = options?.assertions ?? {}
  const helpers = options?.allowHelpers ?? []

  const vocabulary = (family) => [
    ...DEFAULT_ASSERTIONS[family],
    ...(extra[family] ?? []),
    ...helpers,
  ]

  return defineRule({
    id: 'lanekeep/no-assertionless-test',
    language: ['typescript', 'tsx', 'python', 'go', 'rust'],
    severity: 'error',

    card: {
      message: 'assertionless test',
      remediation:
        'assert an observable outcome — a test no outcome can fail protects nothing; if it legitimately cannot assert, skip it or mark it (t.Skip, #[should_panic])',
      examples: {
        bad: "it('adds', () => {\n  add(1, 2)\n})",
        good: "it('adds', () => {\n  expect(add(1, 2)).toBe(3)\n})",
      },
    },

    // The one gate a multi-token judgment can have: where the tests live. Only set when
    // the config says — rust unit tests conventionally live inline in `src/*.rs`, so a
    // default glob would silently exclude them, and a gate that is wrong is worse than
    // none (the `fileContains` entry in AGENTS.md, transposed to paths).
    gates: tests ? { pathMatches: tests } : {},

    // One query per grammar: each matches a *candidate* test definition and captures the
    // body the handler scans. The candidate is narrowed in the handler — by callee name
    // for typescript, by name for python, by name and parameter for go, by attribute for
    // rust — because the narrowing needs text, which a query cannot compare.
    query: {
      typescript: TS_QUERY,
      tsx: TS_QUERY,
      python: `
        (function_definition
          name: (identifier) @name
          body: (block) @body) @def
      `,
      go: `
        (function_declaration
          name: (identifier) @name
          parameters: (parameter_list) @params
          body: (block) @body) @def
      `,
      rust: `
        (function_item
          name: (identifier) @name
          body: (block) @body) @def
      `,
    },

    check(ctx, m) {
      const family = familyOf(ctx.filePath)

      if (family === 'typescript') {
        const callee = normalize(ctx.text(m.fn))
        if (callee !== 'it' && callee !== 'test') return
        if (asserts(ctx, m.body, CALLS.typescript, vocabulary('typescript'))) return
        ctx.report(m.def, 'test asserts nothing')
        return
      }

      const name = ctx.text(m.name)

      if (family === 'python') {
        if (!name.startsWith('test')) return
        // `assert` is a statement, not a call — the node query is the half of the
        // vocabulary a name list cannot carry.
        if (ctx.querySubtree(m.body, '(assert_statement) @a').length > 0) return
        if (asserts(ctx, m.body, CALLS.python, vocabulary('python'))) return
        ctx.report(m.def, `test '${name}' asserts nothing`)
        return
      }

      if (family === 'go') {
        if (!name.startsWith('Test')) return
        // The parameter is what makes go's convention a convention: `TestHelper(data
        // string)` is a name collision, not a test.
        if (!ctx.text(m.params).includes('testing.T')) return
        if (asserts(ctx, m.body, CALLS.go, EXEMPT_GO)) return
        if (asserts(ctx, m.body, CALLS.go, vocabulary('go'))) return
        ctx.report(m.def, `test '${name}' asserts nothing`)
        return
      }

      if (family === 'rust') {
        const attributes = attributesOf(ctx, m.def)
        if (!attributes.some(isTestAttribute)) return
        if (attributes.some((a) => a.includes('should_panic'))) return
        if (asserts(ctx, m.body, CALLS.rustMacros, vocabulary('rust'))) return
        if (asserts(ctx, m.body, CALLS.rustCalls, vocabulary('rust'))) return
        ctx.report(m.def, `test '${name}' asserts nothing`)
      }
    },
  })
}

/** The typescript grammar's test shape; tsx shares the vocabulary, so both entries use it. */
const TS_QUERY = `
  (call_expression
    function: [(identifier) @fn (member_expression object: (identifier) @fn)]
    arguments: (arguments [
      (arrow_function body: (statement_block) @body)
      (function_expression body: (statement_block) @body)
    ])) @def
`

/** What counts as asserting when nothing is configured, per language family. */
const DEFAULT_ASSERTIONS = {
  typescript: ['expect', 'assert'],
  python: ['pytest.raises', 'self.assert', 'self.fail'],
  go: ['t.Error', 't.Fatal', 't.Fail', 'assert.', 'require.'],
  rust: ['assert', 'debug_assert', 'panic'],
}

/** A skipped go test asserts nothing on purpose. */
const EXEMPT_GO = ['t.Skip']

/** The callee-shaped query for each family's assertion scan. */
const CALLS = {
  typescript: '(call_expression function: [(identifier) (member_expression)] @callee) @c',
  python: '(call function: [(identifier) (attribute)] @callee) @c',
  go: '(call_expression function: [(identifier) (selector_expression)] @callee) @c',
  rustMacros: '(macro_invocation macro: [(identifier) (scoped_identifier)] @callee) @c',
  rustCalls:
    '(call_expression function: [(identifier) (scoped_identifier) (field_expression)] @callee) @c',
}

/** Whether any call in `body` has a callee one of `names` prefixes. */
function asserts(ctx, body, query, names) {
  for (const match of ctx.querySubtree(body, query)) {
    const callee = normalize(ctx.text(match.callee))
    if (names.some((name) => callee.startsWith(name))) return true
  }
  return false
}

/** Whitespace stripped, `?.` folded to `.` — the same normalization the callee rules use. */
function normalize(text) {
  return text.replace(/\s+/g, '').replace(/\?\./g, '.')
}

/** Which language family a subject file belongs to, from its extension. */
function familyOf(path) {
  const extension = path.slice(path.lastIndexOf('.') + 1)
  if (extension === 'py') return 'python'
  if (extension === 'go') return 'go'
  if (extension === 'rs') return 'rust'
  return 'typescript'
}

/**
 * The attribute texts sitting directly above a rust item, normalized.
 *
 * Attributes are *sibling* `attribute_item` nodes preceding the `function_item`, so this
 * walks backwards from the item through its parent's children, stepping over comments.
 * The handle comparison is `=== undefined` on purpose — a node handle is an integer and
 * the root's is `0`, so a truthiness check would discard a real node.
 */
function attributesOf(ctx, def) {
  const parent = ctx.parent(def)
  if (parent === undefined) return []

  const siblings = ctx.children(parent)
  const at = siblings.indexOf(def)
  const found = []
  for (let i = at - 1; i >= 0; i -= 1) {
    const kind = ctx.kind(siblings[i])
    if (kind === 'line_comment' || kind === 'block_comment') continue
    if (kind !== 'attribute_item') break
    found.push(normalize(ctx.text(siblings[i])))
  }
  return found
}

/** `#[test]`, or a path attribute whose last segment is `test` — `#[tokio::test]`. */
function isTestAttribute(text) {
  return text === '#[test]' || text.endsWith('::test]')
}

File: crates/lanekeep-rules/tests/no_assertionless_test.rs
//! `lanekeep/no-assertionless-test`, run through the real engine.
//!
//! One rule, four language families: what each language's cases assert is the pair the rule
//! is made of — its test-definition detection and its assertion vocabulary — plus the
//! exemptions that are correctness rather than convenience (`t.Skip`, `#[should_panic]`).
//! The subject file the harness writes is at `subject/input.<ext>`, which is what the
//! `tests` globs here are written against.

#![expect(
    clippy::expect_used,
    reason = "`clippy.toml`'s allow-*-in-tests only reaches `#[test]` functions and \
              `#[cfg(test)]` modules. The helpers below are neither, so the grant it \
              already makes for unit tests has to be restated for them."
)]

use lanekeep_testkit::RuleTester;

fn tester_for(extension: &str, options: &str) -> RuleTester {
    let source = lanekeep_rules::source("no-assertionless-test").expect("the rule ships");
    RuleTester::configured_with_extension("no-assertionless-test", source, extension, options)
        .expect("builds")
        .with_builtins(lanekeep_rules::source)
}

// --- typescript ---------------------------------------------------------------------------

#[test]
fn a_typescript_test_without_an_assertion_is_reported() {
    tester_for("ts", "{}")
        .reports_at("it('adds', () => {\n  add(1, 2)\n})\n", &[(1, 1)])
        .expect("a test body that checks nothing is the failure this rule exists for");
}

#[test]
fn a_typescript_test_with_an_expect_is_fine() {
    tester_for("ts", "{}")
        .accepts("it('adds', () => {\n  expect(add(1, 2)).toBe(3)\n})\n")
        .expect("expect() is the default vocabulary");
}

#[test]
fn the_test_name_and_the_only_variant_are_covered() {
    tester_for("ts", "{}")
        .reports_at(
            "test('adds', function () {\n  add(1, 2)\n})\nit.only('subtracts', () => {\n  sub(3, 1)\n})\n",
            &[(1, 1), (4, 1)],
        )
        .expect("`test(...)` and `it.only(...)` are tests too");
}

#[test]
fn an_ordinary_function_call_is_not_a_test() {
    tester_for("ts", "{}")
        .accepts("setup('adds', () => {\n  add(1, 2)\n})\n")
        .expect("only the test vocabulary defines a test");
}

#[test]
fn a_configured_assertion_name_is_honored() {
    // The ignored-options trap: an ignored vocabulary only ever adds violations, so the
    // accepting direction is what proves the option reached the rule.
    tester_for("ts", "{ assertions: { typescript: ['verify'] } }")
        .accepts("it('adds', () => {\n  verify(add(1, 2))\n})\n")
        .expect("the configured vocabulary counts as asserting");
}

#[test]
fn an_allowed_helper_counts_in_every_language() {
    tester_for("ts", "{ allowHelpers: ['expectValidResponse'] }")
        .accepts("it('responds', () => {\n  expectValidResponse(call())\n})\n")
        .expect("a helper the config vouches for counts as asserting");
}

// --- python -------------------------------------------------------------------------------

#[test]
fn a_python_test_without_an_assertion_is_reported() {
    tester_for("py", "{}")
        .reports_at("def test_add():\n    helper()\n", &[(1, 1)])
        .expect("a test_ function that checks nothing is reported at its definition");
}

#[test]
fn a_python_assert_statement_is_an_assertion() {
    // `assert` is a *statement* in python, not a call — the vocabulary has to be
    // node-shaped there, not only name-shaped.
    tester_for("py", "{}")
        .accepts("def test_add():\n    assert add(1, 2) == 3\n")
        .expect("the assert statement is the language's own assertion form");
}

#[test]
fn python_unittest_methods_and_pytest_raises_are_assertions() {
    tester_for("py", "{}")
        .accepts(
            "class TestAdd(TestCase):\n    def test_add(self):\n        self.assertEqual(add(1, 2), 3)\n\ndef test_raises():\n    with pytest.raises(ValueError):\n        add(None, None)\n",
        )
        .expect("`self.assert*` and `pytest.raises` are the default vocabulary");
}

#[test]
fn an_ordinary_python_function_is_not_a_test() {
    tester_for("py", "{}")
        .accepts("def helper():\n    do_things()\n")
        .expect("a non-test function asserting nothing is fine");
}

#[test]
fn the_tests_globs_gate_where_the_rule_looks() {
    // Both directions, or a gate that is ignored looks exactly like a gate that matches
    // everything.
    let inside = tester_for("py", "{ tests: ['subject/*'] }");
    inside
        .reports_at("def test_add():\n    helper()\n", &[(1, 1)])
        .expect("the subject is inside the tests globs");

    let outside = tester_for("py", "{ tests: ['elsewhere/*'] }");
    outside
        .accepts("def test_add():\n    helper()\n")
        .expect("the subject is outside the tests globs, so the rule never looks");
}

// --- go -----------------------------------------------------------------------------------

#[test]
fn a_go_test_without_an_assertion_is_reported() {
    tester_for("go", "{}")
        .reports_at(
            "package main\n\nimport \"testing\"\n\nfunc TestAdd(t *testing.T) {\n\thelper()\n}\n",
            &[(5, 1)],
        )
        .expect("a Test function that checks nothing is reported at its definition");
}

#[test]
fn go_t_error_and_testify_are_assertions() {
    tester_for("go", "{}")
        .accepts(
            "package main\n\nimport \"testing\"\n\nfunc TestAdd(t *testing.T) {\n\tif add(1, 2) != 3 {\n\t\tt.Errorf(\"wrong\")\n\t}\n}\n\nfunc TestSub(t *testing.T) {\n\trequire.NoError(t, sub())\n}\n",
        )
        .expect("`t.Error*` and `require.*` are the default vocabulary");
}

#[test]
fn a_go_function_without_the_testing_parameter_is_not_a_test() {
    // `TestHelper(data string)` is a name collision, not a test — the `*testing.T`
    // parameter is what makes go's convention a convention.
    tester_for("go", "{}")
        .accepts("package main\n\nfunc TestHelper(data string) {\n\thelper(data)\n}\n")
        .expect("the Test prefix alone does not make a test");
}

#[test]
fn a_skipped_go_test_is_exempt() {
    // A skipped test legitimately asserts nothing; reporting it would punish the honest
    // spelling of "this cannot run here".
    tester_for("go", "{}")
        .accepts(
            "package main\n\nimport \"testing\"\n\nfunc TestAdd(t *testing.T) {\n\tt.Skip(\"needs a database\")\n}\n",
        )
        .expect("t.Skip is an exemption, not an assertion");
}

// --- rust ---------------------------------------------------------------------------------

#[test]
fn a_rust_test_without_an_assertion_is_reported() {
    tester_for("rs", "{}")
        .reports_at("#[test]\nfn adds() {\n    helper();\n}\n", &[(2, 1)])
        .expect("a #[test] function that checks nothing is reported at its definition");
}

#[test]
fn rust_assert_macros_are_assertions() {
    tester_for("rs", "{}")
        .accepts("#[test]\nfn adds() {\n    assert_eq!(add(1, 2), 3);\n}\n")
        .expect("assert!/assert_eq!/assert_ne! are the default vocabulary");
}

#[test]
fn a_should_panic_rust_test_is_exempt() {
    tester_for("rs", "{}")
        .accepts("#[test]\n#[should_panic]\nfn overflows() {\n    add(i32::MAX, 1);\n}\n")
        .expect("the panic is the assertion");
}

#[test]
fn a_function_without_the_test_attribute_is_not_a_test() {
    // `#[cfg(test)]` gates compilation and `#[inline]` is unrelated; neither makes the
    // function a test, so neither may cause a report.
    tester_for("rs", "{}")
        .accepts(
            "#[cfg(test)]\nfn helper() {\n    do_things();\n}\n\nfn plain() {\n    more();\n}\n",
        )
        .expect("only the test attribute defines a test");
}

#[test]
fn an_attribute_path_ending_in_test_is_a_test() {
    // `#[tokio::test]` and friends: the attribute is a path whose last segment is `test`.
    tester_for("rs", "{}")
        .reports_at(
            "#[tokio::test]\nasync fn responds() {\n    call().await;\n}\n",
            &[(2, 1)],
        )
        .expect("a ::test attribute defines a test exactly as #[test] does");
}
